parachain/pallets/basic-channel/src/outbound/mod.rs

```diff
@@ -202,52 +214,65 @@ pub mod pallet {
 				return Err(Error::<T>::Overflow.into());
 			}
 
-			<MessageQueue<T>>::try_append(Message {
-				id: next_id,
-				target,
-				payload: payload.to_vec().try_into().map_err(|_| Error::<T>::PayloadTooLarge)?,
-			})
-			.map_err(|_| Error::<T>::QueueSizeLimitReached)?;
-			Self::deposit_event(Event::MessageAccepted(next_id));
+			<MessageQueue<T>>::try_append(EnqueuedMessage(*who, Message {
+					id: next_id,
+					target,
+					payload: payload.to_vec().try_into().map_err(|_| Error::<T>::PayloadTooLarge)?,
+				}))
+				.map_err(|_| Error::<T>::QueueSizeLimitReached)?;
+ 			Self::deposit_event(Event::MessageAccepted(next_id));
 
 			<NextId<T>>::put(next_id + 1);
 
 			Ok(())
 		}
 
 		fn commit() -> Weight {
+			let who = Self::principal().unwrap();
+			let w2 = who.clone();
+
+			// for every account id in the message queues map, create an Eth ABI-encoded message bundle
+			// these encoded bundles will be the leaves of a merkle tree
 			let messages = <MessageQueue<T>>::take();
```

Comment:
After reading the queued messages, you could convert it into an interim (BTreeMap)[https://crates.parity.io/sp_std/collections/btree_map/struct.BTreeMap.html], where the keys are the AccountId in the enqueued messages, and the values are all the collected messages for that that account.

Then that interim BTreeMap gets converted into a merkle tree.